examples/constraints.rs

```diff
@@ -0,0 +1,217 @@
+use std::{error::Error, io};
+
+use crossterm::{
+    event::{self, DisableMouseCapture, EnableMouseCapture, Event, KeyCode},
+    execute,
+    terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
+};
+use itertools::Itertools;
+use ratatui::{layout::Constraint::*, prelude::*, widgets::*};
+
+fn main() -> Result<(), Box<dyn Error>> {
+    // setup terminal
+    enable_raw_mode()?;
+    let mut stdout = io::stdout();
+    execute!(stdout, EnterAlternateScreen, EnableMouseCapture)?;
+    let backend = CrosstermBackend::new(stdout);
+    let mut terminal = Terminal::new(backend)?;
+
+    // create app and run it
+    let res = run_app(&mut terminal);
+
+    // restore terminal
+    disable_raw_mode()?;
+    execute!(
+        terminal.backend_mut(),
+        LeaveAlternateScreen,
+        DisableMouseCapture
+    )?;
+    terminal.show_cursor()?;
+
+    if let Err(err) = res {
+        println!("{err:?}");
+    }
+
+    Ok(())
+}
+
+fn run_app<B: Backend>(terminal: &mut Terminal<B>) -> io::Result<()> {
+    let mut selection = 0;
+    loop {
+        terminal.draw(|f| ui(f, selection))?;
+
+        if let Event::Key(key) = event::read()? {
+            use KeyCode::*;
+            match key.code {
+                Char('q') => break Ok(()),
+                Char('j') | Char('l') | Down | Right => {
+                    selection = selection.saturating_add(1).min(6)
+                }
+                Char('k') | Char('h') | Up | Left => selection = selection.saturating_sub(1),
+                _ => (),
+            }
+        }
+    }
+}
+
+struct Example {
+    constraints: Vec<Constraint>,
+}
+
+impl Example {
+    fn new<C>(constraints: C) -> Self
+    where
+        C: Into<Vec<Constraint>>,
+    {
+        Self {
+            constraints: constraints.into(),
+        }
+    }
+}
+
+impl Widget for Example {
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        let [title, legend, area] = area.split(&Layout::vertical([Ratio(1, 3); 3]));
+        let heading = Line::from(
+            Itertools::intersperse(
+                self.constraints.iter().enumerate().map(|(i, c)| {
+                    Span::styled(
+                        format!("{:?}", c),
+                        Style::default().fg(Color::Indexed(i as u8 + 1)),
+                    )
+                }),
+                Span::raw(", "),
+            )
+            .collect::<Vec<Span>>(),
+        );
+        Paragraph::new(heading)
+            .block(Block::default().padding(Padding::new(0, 0, 1, 0)))
```

Comment:
Why pad this - it's a single line right?
Give it an area 2 high to write into in the layout.

Reply:
If I remove the padding I get this:

<img width="580" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/8a3a6c78-c6f0-495f-8415-79ee5a1d8ce7">

I think this looks nicer:

<img width="576" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/6ad8ad94-9059-4ccf-8c34-189b4ec5dabf">

It's a padding from the top, here it is with padding `2`:

<img width="576" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/9386792e-f464-4812-b7fb-112595628b4c">

I changed it to `Padding::vertical(1)` so it is clearer from the code.

Related:

Should `Padding` get some additional functions?

Instead of this:

```
Block::default().padding(Padding::new(0, 0, 2, 0))
```

It should be:

```
Block::default().padding(Padding::top(2))
```

Reply:
Ahh my confusion was that this was bottom padding 😀 
Good idea about adding better constructors

Reply:
https://github.com/ratatui-org/ratatui/issues/798